server/expression/in_subquery.go

```diff
@@ -0,0 +1,259 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/types"
+	vitess "github.com/dolthub/vitess/go/vt/sqlparser"
+
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// InSubquery represents a VALUE IN (SELECT ...) expression.
+type InSubquery struct {
+	leftExpr  sql.Expression
+	rightExpr *plan.Subquery
+
+	// These variables are used so that we can resolve the comparison functions once and reuse them as we iterate over rows.
+	// These are assigned in WithChildren, so refer there for more information.
+	leftLiteral   *Literal
+	rightLiterals []*Literal
+	compFuncs     []*framework.CompiledFunction
+}
+
+var _ vitess.Injectable = (*InSubquery)(nil)
+var _ sql.Expression = (*InSubquery)(nil)
+var _ expression.BinaryExpression = (*InSubquery)(nil)
+
+// NewInSubquery returns a new *InSubquery.
+func NewInSubquery() *InSubquery {
+	return &InSubquery{}
+}
+
+// Children implements the sql.Expression interface.
+func (in *InSubquery) Children() []sql.Expression {
+	return []sql.Expression{in.leftExpr, in.rightExpr}
+}
+
+var nilKey, _ = sql.HashOf(sql.NewRow(nil))
+
+// Eval implements the sql.Expression interface.
+func (in *InSubquery) Eval(ctx *sql.Context, row sql.Row) (any, error) {
+	if len(in.compFuncs) == 0 {
+		return nil, fmt.Errorf("%T: cannot Eval as it has not been fully resolved", in)
+	}
+
+	left, err := in.leftExpr.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// The NULL handling for IN expressions is tricky. According to
+	// https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_in:
+	// To comply with the SQL standard, IN() returns NULL not only if the expression on the left hand side is NULL, but
+	// also if no match is found in the list and one of the expressions in the list is NULL.
+	// However, there's a strange edge case. NULL IN (empty list) return 0, not NULL.
+	leftNull := left == nil
+
+	if types.NumColumns(in.Left().Type()) != types.NumColumns(in.Right().Type()) {
+		return nil, sql.ErrInvalidOperandColumns.New(types.NumColumns(in.Left().Type()), types.NumColumns(in.Right().Type()))
+	}
+
+	typ := in.rightExpr.Type()
+	right := in.rightExpr
+
+	// TODO: does this work for all pg values?
+	values, err := right.HashMultiple(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// NULL IN (list) returns NULL. NULL IN (empty list) returns 0
+	if leftNull {
+		if values.Size() == 0 {
+			return false, nil
+		}
+		return nil, nil
+	}
+
+	// convert left to right's type
+	nLeft, _, err := typ.Convert(left)
```

Comment:
We don't use the `Convert` functions, that's a relic of GMS that we can't remove since we must adhere to the interface. This won't do anything, and will just fail if the types are different (also means that valid queries will fail because of this). Postgres uses casts for conversions. But from looking at the documentation, it doesn't look like we're supposed to cast anything anyway.
https://www.postgresql.org/docs/15/functions-subquery.html
> The left-hand expression is evaluated and compared to each row of the subquery result. The result of IN is “true” if any equal subquery row is found. The result is “false” if no equal row is found (including the case where the subquery returns no rows).

This implies that we're using the equality operator (`=`), which will implicitly cast if it's valid and viable.

Reply:
This was harmless but useless, removed.